concurrent_test.go

```diff
@@ -292,138 +236,212 @@ func runWorker(t *testing.T, w worker, errCh chan error) (historyRecords, error)
 	return rs, err
 }
 
-type worker interface {
-	name() string
-	run() (historyRecords, error)
-}
-
-type readWorker struct {
+type worker struct {
 	id int
 	db *btesting.DB
 
 	bucket []byte
 	keys   []string
 
-	readTime duration
+	conf concurrentConfig
 
 	errCh  chan error
 	stopCh chan struct{}
 
 	t *testing.T
 }
 
-func (r *readWorker) name() string {
-	return fmt.Sprintf("readWorker-%d", r.id)
+func (w *worker) name() string {
+	return fmt.Sprintf("worker-%d", w.id)
 }
 
-func (r *readWorker) run() (historyRecords, error) {
+func (w *worker) run() (historyRecords, error) {
 	var rs historyRecords
 	for {
 		select {
-		case <-r.stopCh:
-			r.t.Log("Reading transaction finished.")
+		case <-w.stopCh:
+			w.t.Logf("Worker %q finished.", w.name())
 			return rs, nil
 		default:
 		}
 
-		err := r.db.View(func(tx *bolt.Tx) error {
-			b := tx.Bucket(r.bucket)
-
-			selectedKey := r.keys[mrand.Intn(len(r.keys))]
-			initialVal := b.Get([]byte(selectedKey))
-			time.Sleep(randomDurationInRange(r.readTime.min, r.readTime.max))
-			val := b.Get([]byte(selectedKey))
-
-			if !reflect.DeepEqual(initialVal, val) {
-				return fmt.Errorf("read different values for the same key (%q), value1: %q, value2: %q",
-					selectedKey, formatBytes(initialVal), formatBytes(val))
-			}
-
-			clonedVal := make([]byte, len(val))
-			copy(clonedVal, val)
+		op := w.randomOperation()
+		rec, err := op.run()
+		if err != nil {
+			readErr := fmt.Errorf("[%s: %s]: %w", w.name(), op.name(), err)
+			w.t.Error(readErr)
+			w.errCh <- readErr
+			return rs, readErr
+		}
 
-			rs = append(rs, historyRecord{
-				OperationType: Read,
-				Key:           selectedKey,
-				Value:         clonedVal,
-				Txid:          tx.ID(),
-			})
+		rs = append(rs, rec)
+		time.Sleep(randomDurationInRange(w.conf.workInterval.min, w.conf.workInterval.max))
+	}
+}
 
-			return nil
-		})
+func (w *worker) pickOperation() OperationType {
+	sum := 0
+	for _, op := range w.conf.operationRatio {
+		sum += op.chance
+	}
+	roll := mrand.Int() % sum
+	for _, op := range w.conf.operationRatio {
+		if roll < op.chance {
+			return op.operation
+		}
+		roll -= op.chance
+	}
+	panic("unexpected")
+}
 
-		if err != nil {
-			readErr := fmt.Errorf("[reader error]: %w", err)
-			r.t.Error(readErr)
-			r.errCh <- readErr
-			return rs, readErr
+func (w *worker) randomOperation() operation {
+	op := w.pickOperation()
+	switch op {
+	case Read:
+		return &readOperation{
+			db:           w.db,
+			bucket:       w.bucket,
+			keys:         w.keys,
+			readInterval: w.conf.readInterval,
+		}
+	case Write:
+		return &writeOperation{
+			db:         w.db,
+			bucket:     w.bucket,
+			keys:       w.keys,
+			writeBytes: w.conf.writeBytes,
+		}
+	case Delete:
+		return &deleteOperation{
+			db:     w.db,
+			bucket: w.bucket,
+			keys:   w.keys,
 		}
 	}
+	panic("unexpected")
 }
 
-type writeWorker struct {
-	id int
-	db *btesting.DB
+type operation interface {
+	name() string
+	run() (historyRecord, error)
+}
```

Comment:
Having operation interface for internal package doesn't seem appropriate. It makes no sense to have interface with just methods not taking any arguments. This can be easily expressed as a switch statement in worker that calls function `read(db, bucket, keys)`, `write`, `delete`. 
It will result in less and more readable code.

Reply:
Please read line `297 ~ 320.` 

> Having operation interface for internal package doesn't seem appropriate. It makes no sense to have interface with just methods not taking any arguments.

Do not agree with this. `Generic worker + operation interface` seems good to me. Do not want to get all details included in worker.

Reply:
This is incorrect way to model the problem. I agree that worker should not be aware of detail of how each operation functions, basically following the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) here is good. However, you are incorrectly using the [Strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern).

You create a fake basically ephemeral structs for operations, put all context to them running execute. This has two problems:
* You are modeling operation type for second time. first time is type alias `type OperationType string`. Second time is operation struct for each operation. First one is modeling via data, second via type. I would suggest using only the first one.
* You unnecessarily complicating process of executing operation. For me operation should like any other function, it takes takes all arguments it needs and then it can execute. By doing it via struct you are hiding the real arguments needed to execute operation and add another unnecessary step to create the struct.

My suggestion would be to just use `type OperationType string` and have a function `runOperation(OperationType)`.

Like:
```
func runOperation(op OperationType, db *btesting.DB, bucket []byte, keys []string, config *concurrentConfig) (historyRecord, err) {
  switch op {
  case Read:
    return read(db, bucket, keys, config.readInterval duration)
  ...
  default:
    panic("unknown operation type")
  }
}

func read(db *btesting.DB, bucket []byte, keys []string, readInterval ) (historyRecord, err) {
...
}
```

Reply:
thx & resolved.